site/src/pages/CreateWorkspacePage/CreateWorkspacePage.tsx

```diff
@@ -33,39 +38,141 @@ const CreateWorkspacePage: FC = () => {
-  const [createWorkspaceState, send] = useMachine(createWorkspaceMachine, {
-    context: {
-      organizationId,
-      templateName,
-      mode,
-      defaultBuildParameters,
-      defaultName:
-        mode === "auto" ? generateUniqueName() : searchParams.get("name") ?? "",
-      versionId: searchParams.get("version") ?? undefined,
-    },
-    actions: {
-      onCreateWorkspace: (_, event) => {
-        navigate(`/@${event.data.owner_name}/${event.data.name}`);
-      },
-    },
+  const customVersionId = searchParams.get("version") ?? undefined;
+  const defaultName =
+    mode === "auto" ? generateUniqueName() : searchParams.get("name") ?? "";
+
+  const queryClient = useQueryClient();
+  const autoCreateWorkspaceMutation = useMutation(
+    autoCreateWorkspace(queryClient),
+  );
+  const createWorkspaceMutation = useMutation(createWorkspace(queryClient));
+
+  const templateQuery = useQuery(templateByName(organizationId, templateName));
+  const permissionsQuery = useQuery(
+    checkAuthorization({
+      checks: createWorkspaceChecks(organizationId),
+    }),
+  );
+  const realizedVersionId =
+    customVersionId ?? templateQuery.data?.active_version_id;
+  const richParametersQuery = useQuery({
+    ...richParameters(realizedVersionId ?? ""),
+    enabled: realizedVersionId !== undefined,
   });
-  const { template, parameters, permissions, defaultName, versionId } =
-    createWorkspaceState.context;
-  const title = createWorkspaceState.matches("autoCreating")
+  const realizedParameters = richParametersQuery.data
+    ? richParametersQuery.data.filter(paramsUsedToCreateWorkspace)
+    : undefined;
+
+  const { externalAuth, externalAuthPollingState, startPollingExternalAuth } =
+    useExternalAuth(realizedVersionId);
+
+  const isLoadingFormData =
+    templateQuery.isLoading ||
+    permissionsQuery.isLoading ||
+    richParametersQuery.isLoading;
+  const loadFormDataError =
+    templateQuery.error ?? permissionsQuery.error ?? richParametersQuery.error;
+
+  const title = autoCreateWorkspaceMutation.isLoading
     ? "Creating workspace..."
     : "Create workspace";
 
+  const onCreateWorkspace = useCallback(
+    (workspace: Workspace) => {
+      navigate(`/@${workspace.owner_name}/${workspace.name}`);
+    },
+    [navigate],
+  );
+
+  useEffect(() => {
+    if (mode === "auto") {
+      autoCreateWorkspaceMutation
+        .mutateAsync({
+          templateName,
+          organizationId,
+          defaultBuildParameters,
+          defaultName,
+          versionId: realizedVersionId,
+        })
+        .then(onCreateWorkspace)
+        .catch(() => {
+          searchParams.delete("mode");
+          setSearchParams(searchParams);
+        });
+    }
+  }, [
+    autoCreateWorkspaceMutation,
+    defaultBuildParameters,
+    defaultName,
+    mode,
+    onCreateWorkspace,
+    organizationId,
+    realizedVersionId,
+    searchParams,
+    setSearchParams,
+    templateName,
```

Comment:
I feel like this might be a good use case for the `useEffectEvent` polyfill. If I'm reading this right, that would let you simplify the dependency array to just the function from that hook, plus the `mode` variable